Initial Javascript bindings work

JoshuaGross opened the pull request:
Javascript bindings including unit tests, streaming interface (the `dispatch` method). I can provide some demo tomorrow and get the tests running on Travis.

mookerji commented:
!!!

gsmcmullin commented:
Can we please break the anti-pattern of keeping the generated files in version control?

JoshuaGross commented:
@gsmcmullin Why don't you like checking in generated code? In my experience it makes debugging (especially things like `bisect`) significantly easier as you don't need to consider if someone had a build environment set up differently or go through any extra steps besides checking out the code. Are you concerned about history being "messy"?

mfine commented:
Checking in the code is ok. This repo serves both the purpose of generating code and hosting code artifacts (with some non generated artifacts as well). It's a common pattern in generated libraries. See:

https://github.com/brendanhay/amazonka
https://github.com/brendanhay/gogol
https://github.com/aws/aws-sdk-go/tree/master/service
https://github.com/boto/boto/tree/develop/boto

gsmcmullin commented:
It bothers me because I can't review a diff that's 90% generated crap.  Because others do it doesn't make it a good idea.

JoshuaGross commented:
If that's the only concern we can easily split the generated-code-only changes into their own commits so that they can be isolated. That said, that code needs to be reviewed anyway.

mookerji commented on javascript/sbp/acquisition.js:
```diff
@@ -0,0 +1,177 @@
+ratio.
+
+
+  Parameters
```
The comment format here is for numpy's Doxygen comments. If there's some kind of autogenerated documentation for javascript using embedded comment annotations, I recommend using that.

mookerji replied:
Like I think the comment would go above the function/object closure like maybe: 

```
/**
 * SBP ...
 *
 * @param {Object} options
 * @access public
 */
```

mookerji replied:
Also, fixing some of the spacing and alignment will definitely make the source a bit easier to read.

mookerji replied:
nit: Ok, the alignment comment was bullshit. Crunching the spacing down in the comments would be cool.

jacobmcnamee commented:
Can we move the generated files to a separate repo? Having everything in one place makes code review tedious, and it's very easy for generated files to get out of sync.

I don't see the use case for needing to download both the generated files and the build system, let alone in the same repo.

mfine commented:
From a customer getting tripped up on missing generated Java files...

```Hello! 

I'm trying to send Piksi data to an Android device via Bluetooth. I'm using your libsbp library to decode the messages. 

Is it possible that there are some files missing in you github repository? The example uses the class MsgTrackingState but I can't find this or any other message classes in the java folder. However, this classes exist in the python folder. 

I tried to implement the class MsgPosLLH myself, but I can't get it to work properly. 

Kind regards, 
David```

mookerji commented:
@gsmcmullin @jacobmcnamee, so in every project of this kind I've seen, it's been typical to keep generated code and the compiler under version control in the same repository. I was going to list examples but it seems @mfine has already done that. I disagree pretty strongly about splitting this repository up or keeping the source code out of version control. If this is an antipattern, I'd prefer to see some more substantive arguments describing why its necessary rather than the statement that others are Wrong.

Some points:
* Having multiple repositories or separate release artifacts introduces more management overhead to keep track off than we currently have. If you think it's tedious now to code review or manage multi-language clients across directories, just wait until they're split into multiple repositories. If we generated the source code as release artifacts, most users and developers would likely download and review those separately. At this point, you'd go through a more roundabout approach to performing what we currently do. The easy thing with the current repository structure is a very flat information hierarchy: you can read through source, review it, and deploy it all from a single place. It's tedious, yes, but typically for large commits and less arguably tedious than the alternatives.
* The primary artifact here for all consumers is the generated source code. How that source code is generated is different for each client, but ultimately it's intended to be human readable and reviewed. You cannot review the generator of the code and then reason, "Yes! This AST will generate the source I want.". It is the generated source code that is meant to be reviewed and consumed by our clients, which is why it's checked into version control.
* If the generated files are out-of-sync, then we can put a procedure (git commit hook, etc.) into place that fixes that, because it definitely sounds like a problem. I totally agree.
* The time spent with repository gymnastics and yakshaving is time best spent elsewhere.

mookerji commented on generator/sbpg/targets/javascript.py:
```diff
@@ -0,0 +1,155 @@
+  'double': 8
+}
+
+PYDOC_CODE = {
```
not pydoc. Also does js have different numeric types?

JoshuaGross replied:
Unfortunately not, everything is `number`. Numbers can have decimal points, or not.

mookerji replied:
Yeah, you're probably going to want to change that in generated comments.

JoshuaGross replied:
Done. I changed to the actual JavaScript type but with more detail, which I think folks will find useful. Example: `u16` now maps to `number (unsigned 16-bit int, 2 bytes)`

mookerji commented on javascript/README.md:
```diff
@@ -0,0 +1,13 @@
+
+From the parent directory, run `make javascript` to regenerate bindings.
+
+# Using the JavaScript libsbp bindings
```
Not really related to this readme, but can we package the javascript client as well? npm or something?

JoshuaGross replied:
yeah, I intend to do that - hopefully as part of the build process - once this PR has been approved

mookerji replied:
Lovely

mookerji commented on javascript/sbp/acquisition.js:
```diff
@@ -0,0 +1,148 @@
+MSG_ACQ_RESULT_DEP_A.prototype = Object.create(SBP.prototype);
+MSG_ACQ_RESULT_DEP_A.prototype.constructor = MSG_ACQ_RESULT_DEP_A;
+MSG_ACQ_RESULT_DEP_A.prototype.toBuffer = function () {
+  var buffers = [];
```
Hmmm, so this inlining of allocating all these buffers and concatenating them seems a bit tedious. Also, does `binary-parser` not have a method for binary serializing?

JoshuaGross replied:
It appears that they don't. I'm happy to make this `toBuffer` routine less tedious-looking. I already moved some of the stuff to `sbp.js` (it used to all be inlined).

mookerji replied:
Wow, that's really fucking weird.

Uh, yeah a utility function or something might be DRY here.  (or maybe submitting a PR to `binary-parser` for that method). For what it's worth though, the use case of actually materializing a buffer from a JSON object here is likely to be rare (in the short term)? Typically, that's something you'd want to do if you were writing commands over to the Piksi (like from the console), so perhaps it's not actually worth worrying about too much.

mookerji commented on javascript/sbp/observation.js:
```diff
@@ -0,0 +1,1016 @@
+MSG_OBS.prototype.parser = new Parser()
+  .endianess('little')
+  .nest('header', { type: ObservationHeader.prototype.parser })
+  .array('obs', { type: PackedObsContent.prototype.parser, readUntil: 'eof' });
```
eof?

The observation messages can definitely be super-complicated to get right, from the declarative parsing point of view. Did this work? The tracking message is similar.

JoshuaGross replied:
It's passing the tests so I guess so? `eof` is end-of-file, is that not what's expected? Basically the payload is parsed separately (as in Haskell and Python implementations AFAIK) so we can read to the end of the payload which reads until (in effect) immediately before the crc.

mookerji replied:
I was confused for a minute. If `eof` == `end of buffer`, that's fine.

JoshuaGross replied:
yeah, exactly

mookerji commented on python/sbp/settings.py:
```diff
@@ -300,7 +300,7 @@ def to_json_dict(self):
   """SBP class for message MSG_SETTINGS_READ_BY_INDEX_REQ (0x00A2).
 
-  You can have MSG_SETTINGS_READ_BY_INDEX_REQ inherent its fields directly
+  You can have MSG_SETTINGS_READ_BY_INDEX_REQ inherit its fields directly
```
Eh, yeah, it probably would have made more sense to keep the generated python stuff with this comment spelling correction into a separate commit.

JoshuaGross replied:
word. I can pull this out

mookerji commented on javascript/sbp/observation.js:
```diff
@@ -0,0 +1,1016 @@
+  .uint16('lock')
+  .uint8('prn');
+
+var MSG_OBS_DEP_A = function (sbp, msg) {
```
These object names should probably be cased (vs capped). All caps has typically been used for constants.

JoshuaGross replied:
will do

JoshuaGross replied:
done

mookerji commented on javascript/sbp/observation.js:
```diff
@@ -0,0 +1,1016 @@
+  .uint16('lock')
+  .uint32('sid');
+
+var MSG_OBS = function (sbp, msg) {
```
Hrmmmm, what's both `sbp` and `msg` doing here? Is `sbp` a mixin?

You probably want to keep the original calling contract (using an existing sbp object or instantiating an object from a dict of fields):

```python
def __init__(self, sbp=None, **kwargs):
    if sbp:
      super( MsgTrackingIq,
             self).__init__(sbp.msg_type, sbp.sender, sbp.length,
                            sbp.payload, sbp.crc)
      self.from_binary(sbp.payload)
```

JoshuaGross replied:
`msg` should be renamed to `payload` or I can just use `sbp.payload`

JoshuaGross commented:
@mookerji I replaced the inlined `toBuffer` code generation with `fieldSpecs`.

mookerji commented on javascript/sbp/piksi.js:
```diff
@@ -0,0 +1,366 @@
+/**
+ * SBP class for message MSG_ALMANAC (0x0069).
+ *
+ * This is a legacy message for sending and loading a satellitealamanac onto the Piksi's flash memory from the host.
```
word wrap this description and the field comments as well?

mookerji commented on javascript/sbp/acquisition.js:
```diff
@@ -0,0 +1,93 @@
+ * This message describes the results from an attempted GPS signalacquisition search for a satellite PRN over a code phase/carrierfrequency range. It contains the parameters of the point in theacquisition search space with the best signal-to-noise (SNR)ratio.
+ *
+ * Fields in the SBP payload (`sbp.payload`):
+ * @field snr number (float, 4 bytes) SNR of best point. Currently dimensonless, but will haveunits of dB Hz in the revision of this message.
```
nit: word wrap on comment and spacing (`haveunits`).

mookerji commented on javascript/sbp/msg.js:
```diff
@@ -0,0 +1,210 @@
+  decode: function decode (msg) {
+    var sbp = parser.parse(msg);
+
+    return new sbpTable[sbp['msg_type']](sbp);
```
Hey, so what happens here if the dispatch fails because the `msg_type` is missing? I'm assuming you get some kind of dictionary-related exception/error. In the past, we've caught these, logged a warning, and passed the raw SBP message through.

https://github.com/swift-nav/libsbp/blob/master/python/sbp/table.py#L68

JoshuaGross replied:
Great, good call. Fixing this and adding a test in the next commit.

mookerji commented on javascript/sbp/file_io.js:
```diff
@@ -0,0 +1,240 @@
+/**
+ * SBP class for message MSG_FILEIO_READ_REQ (0x00A8).
+ *
+ * The file read message reads a certain length (up to 255 bytes)from a given offset into a file, and returns the data in aMSG_FILEIO_READ_RESP message where the message length fieldindicates how many bytes were succesfully read.The sequencenumber in the request will be returned in the response.If the message is invalid, a followup MSG_PRINT message willprint "Invalid fileio read message". A devi ...
```
nit: More spacing errors `aMSG_FILEIO_...`

JoshuaGross replied:
resolved - I fixed all the spacing errors (AFAICT). I improved the word wrapping significantly.

mookerji commented on javascript/sbp/msg.js:
```diff
@@ -0,0 +1,210 @@
+      var preamblePos;
+      for (preamblePos = 0; preamblePos < streamBuffer.length; preamblePos++) {
+        preamble = streamBuffer.readUInt8(preamblePos);
+        if (preamble == 0x55) {
```
nit: magic number

JoshuaGross replied:
resolved

JoshuaGross commented:
@mookerji @mfine 

1. Rebased into two commits, one for generation code/scaffolding, and one for the generated code.
2. Removed noisy Python changes that were in previous commits.
3. I cleaned up the generated comments significantly - they're actually readable now.
4. Ponies.

mfine commented:
This is awesome. The 2 commit split works great! Thanks!

JoshuaGross commented:
Fixed to work with NodeJS 0.10.x, 0.12.x, v4.x and v5.x (this is why the Travis build was failing).

mookerji commented:
Ship it!

JoshuaGross commented:
:dancer:
